concent_api/core/payments/report.py

```diff
@@ -0,0 +1,759 @@
+        ('additional_verification_payments_for_provider', list),
+    ]
+)
+
+
+Summary = NamedTuple(
+    'Summary',
+    [
+        ('number_of_pairs', int),
+        ('number_of_subtasks', int),
+        ('number_of_regular_payments', int),
+        ('number_of_forced_payments', int),
+        ('total_amount_to_be_paid', int),
+        ('total_amount_actually_paid', int),
+        ('total_amount_not_paid', int),
+    ]
+)
+
+
+PairReport = NamedTuple(
+    'PairReport',
+    [
+        ('requestor', Client),
+        ('provider', Client),
+        ('pending_responses', PendingResponses),
+        ('payments', Payments),
+        ('errors', list),
+        ('warnings', list),
+        ('summary', Summary),
+    ]
+)
+
+
+Report = NamedTuple(
+    'Report',
+    [
+        ('pair_reports', list),
+        ('summary', Summary),
+    ]
+)
+
+
+Error = NamedTuple(
+    'Error',
+    [
+        ('matching_error', MatchingError),
+        ('message', str),
+    ]
+)
+
+
+Warnings = NamedTuple(
+    'Warnings',
+    [
+        ('matching_warning', MatchingWarning),
+        ('message', str),
+    ]
+)
+
+
+def create_report() -> Report:
+    """ Create payments report for Concent. """
+    pair_reports = []
+
+    for subtask_pair in get_client_pairs_from_subtasks():  # pylint: disable=not-an-iterable
+        related_pending_responses_for_pair = get_related_pending_responses_for_pair(
+            subtask_pair.requestor,
+            subtask_pair.provider,
+        )
+
+        task_to_compute = deserialize_message(subtask_pair.task_to_compute.data.tobytes())
+
+        related_payments_for_pair = get_related_payments_for_pair(
+            task_to_compute.requestor_ethereum_address,
+            task_to_compute.provider_ethereum_address,
+        )
+
+        (errors, warnings) = get_errors_and_warnings(
+            related_payments_for_pair,
+            related_pending_responses_for_pair,
+        )
+
+        pair_report = PairReport(
+            requestor=subtask_pair.requestor,
+            provider=subtask_pair.provider,
+            pending_responses=related_pending_responses_for_pair,
+            payments=related_payments_for_pair,
+            errors=errors,
+            warnings=warnings,
+            summary=create_pair_summary(
+                requestor=subtask_pair.requestor,
+                provider=subtask_pair.provider,
+                pending_responses=related_pending_responses_for_pair,
+                payments=related_payments_for_pair,
+            ),
+        )
+        pair_reports.append(pair_report)
+
+    return Report(
+        pair_reports,
+        summary=create_report_summary(pair_reports),
+    )
+
+
+def get_client_pairs_from_subtasks() -> QuerySet:
+    """
+    Get all unique pairs of Ethereum accounts where one belongs to a provider and the other to the requestor.
+    """
+    return Subtask.objects.distinct(
+        'provider', 'requestor'
+    )
```

Comment:
Right, I though it works differently and pure `distinct(*fields)` is enough. It works correctly for me in both versions, so maybe it have impact when there is greater number of subtasks or they are created in different order.

> Does this code really work? If it does then it returns a random subtask for each pair. This is wrong because the report should process all the available subtasks.

Here I am getting only Clients pairs and TTC. All available subtasks for each pair are retrieved in different place.

:+1: done